language/move-stdlib/nursery/sources/Capability.move

```diff
@@ -0,0 +1,176 @@
+/// scenarios which go beyond having a signer?
+///
+/// The current way how to achieve this in Move is to build a wrapper around `Capability::Cap<T>`.
+/// The wrapper type will be owned by a specific module, restricting how values of it can be created.
+/// Below, we extend the example from above to illustrate this pattern:
+///
+/// ```
+/// struct ProtectedFeatureCap has copy, drop {
+///     cap: Capability::Cap<ProtectedFeature>
+/// }
+///
+/// public fun acquire_protected_feature_access(s: &signer): ProtectedFeatureCap {
+///   let cap = Capability::acquire<ProtectedFeature>(s);
+///   validate_authorization(s, cap); // Do any additional authorization validation
+///   ProtectedFeatureCap{cap}
+/// }
+/// ```
+///
+/// ## Delegation
+///
+/// Capabilities come with the optional feature of *delegation*. Via delegation, an owner of a capability
+/// can designate another signer to be also capable of acquiring the capability. Like the original owner,
+/// the delegate needs to present his signer to obtain the capability in his transactions. Delegation can
+/// be revoked, removing this access right from the delegate.
+///
+/// While the basic authorization mechanism for delegates is the same as with core capabilities, the
+/// target of delegation might be subject of restrictions which need to be specified and verified. This can
+/// be done via global invariants in the specification language. For example, in order to prevent delegation
+/// all together for a capability, one can use the following invariant:
+///
+/// ```
+///   invariant forall a: address where exists<CapState<ProtectedFeature>>(addr):
+///               len(Capability::spec_delegates<ProtectedFeature>(a)) == 0;
+/// ```
+///
+/// Similarly, the following invariant would enforce that delegates, if existent, must satisfy a certain
+/// predicate:
+///
+/// ```
+///   invariant forall a: address where exists<CapState<ProtectedFeature>>(addr):
+///               forall d in Capability::spec_delegates<ProtectedFeature>(a):
+///                  is_valid_delegate_for_protected_feature(d);
+/// ```
+///
+module Std::Capability {
+    use Std::Errors;
+    use Std::Signer;
+    use Std::Vector;
+
+    const ECAP: u64 = 0;
+    const EDELEGATE: u64 = 1;
+
+    /// The token representing an acquired capability. Cannot be stored in memory, but copied and dropped freely.
+    struct Cap<phantom Feature> has copy, drop {
+        root: address
+    }
+
+    /// An internal data structure for representing a configured capability.
+    struct CapState<phantom Feature> has key {
+        delegates: vector<address>
+    }
+
+    /// An internal data structure for representing a configured delegated capability.
+    struct CapDelegateState<phantom Feature> has key {
+        root: address
+    }
+
+    /// Creates a new capability class, owned by the passed signer.
+    public fun create<Feature>(owner: &signer) {
```

Comment:
Should this function take in a `&Feature` as a proof of ownership of the feature? Otherwise what's stopping a malicious user from calling this public function with a protected feature like this:
```
module 0x2::MaliciousModule {
    use Std::Capability;
    use DiemFramework::ProtectedModule::ProtectedFeature;

    public fun steal_cap(malicious_account: &signer) {
        Capability::create<ProtectedFeature>(malicious_account);
    }
}
```

Reply:
Yes, we either need `fun create<private ProtectedFeature>` or we can simulate with a witness (great idea).

Reply:
Updated the code based on your suggestion to have a witness parameter.